Framework/Geometry/src/Instrument/DetectorInfo.cpp

```diff
@@ -228,6 +228,48 @@ DetectorInfo::signedTwoTheta(const std::pair<size_t, size_t> &index) const {
   return angle;
 }
 
+double DetectorInfo::azimuthal(const size_t index) const {
+  if (isMonitor(index))
+    throw std::logic_error("Azimuthal angle is not defined for monitors");
+
+  const auto samplePos = samplePosition();
+  const auto beamLine = samplePos - sourcePosition();
+
+  if (beamLine.nullVector()) {
+    throw Kernel::Exception::InstrumentDefinitionError(
+        "Source and sample are at same position!");
+  }
+
+  const auto sampleDetVec = position(index) - samplePos;
+  const double dotHorizontal = sampleDetVec.scalar_prod(
+      m_instrument->getReferenceFrame()->vecPointingHorizontal());
+  const double dotVertical = sampleDetVec.scalar_prod(
+      m_instrument->getReferenceFrame()->vecPointingUp());
```

Comment:
Shouldn't this projection be done using axes that are perpendicular to the (actual) beam direction? Otherwise I do not understand what this function is supposed to return.

Reply:
I need to improve the docs, but I want to confirm that the math isn’t wrong. The idea is to get the angle from the “horizontal” axis constrained in the horizontal-vertical plane. For “simple” instruments with the beam along z, x in the horizontal, and y vertical, it is just the arctan(y, x). I think the implementation I have does that in the generic case as well.

Reply:
I think your math is ok as long as the beam is exactly in Z direction (or whatever is given by the reference frame). In practice I believe this is often not the case?